src/libraries/System.Private.CoreLib/src/System/Text/ASCIIUtility.cs

```diff
@@ -673,13 +759,35 @@ private static unsafe nuint GetIndexOfFirstNonAsciiChar_Sse2(char* pBuffer, nuin
 
             // Read the first vector unaligned.
 
-            firstVector = Sse2.LoadVector128((ushort*)pBuffer); // unaligned load
+            if (Sse2.IsSupported)
+            {
+                firstVector = Sse2.LoadVector128((ushort*)pBuffer); // unaligned load
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                firstVector = AdvSimd.LoadVector128((ushort*)pBuffer); // unaligned load
+            }
+            else
+            {
+                throw new PlatformNotSupportedException();
+            }
 
             // The operation below forces the 0x8000 bit of each WORD to be set iff the WORD element
             // has value >= 0x0800 (non-ASCII). Then we'll treat the vector as a BYTE vector in order
             // to extract the mask. Reminder: the 0x0080 bit of each WORD should be ignored.
 
-            currentMask = (uint)Sse2.MoveMask(Sse2.AddSaturate(firstVector, asciiMaskForAddSaturate).AsByte());
+            if (Sse2.IsSupported)
+            {
+                currentMask = (ulong)Sse2.MoveMask(Sse2.AddSaturate(firstVector, asciiMaskForAddSaturate).AsByte());
+            }
+            else if (AdvSimd.Arm64.IsSupported)
+            {
+                currentMask = GetMaskOfMostSignificantBits(AdvSimd.AddSaturate(firstVector, asciiMaskForAddSaturate).AsByte());
+            }
+            else
+            {
+                throw new PlatformNotSupportedException();
+            }
 
             if ((currentMask & NonAsciiDataSeenMask) != 0)
```

Comment:
@GrabYourPitchforks: Is NonAsciiDataSeenMask really correct here? Or should it be `0xFFFF`?

Reply:
`0xAAAA` is correct. The way this logic works is that it adds `0x7F80` to every 16-bit `ushort` in the input, saturating at `0xFFFF` (to avoid integer overflow back to a negative value).

This means that if the original value is in the range `0x0000 .. 0x007F` (ASCII), the add operation will result in `0x7F80 .. 0x7FFF`.
If the original value is in the range `0x0080 .. 0x807F` (non-ASCII), the add operation will result in `0x8000 .. 0xFFFF`.
And if the original value is in the range `0x8080 .. 0xFFFF` (non-ASCII), the add operation will result in `0xFFFF` due to saturation.

The non-ASCII case _always_ results in the sum having the `0x8000` bit set, but it doesn't always result in the sum having the `0x0080` bit set. Hence `0xAAAA` - not `0xFFFF` - is the correct comparison against the _pmovmskb_ result.

Reply:
Thanks for the explanation. That makes sense. I didn't realize that `MoveMask()` only takes `byte`, `sbyte` or `double` and here we are operating on `ushort`. Alternative way of doing that operation could be although not sure which amoung `AddSaturate()` and `And()` is cheaper.
```csharp
Vector128<ushort> mask = Vector128.Create((ushort)0xFF80);
var masked = Sse2.And(firstVector, mask).AsSByte();
uint currentMask = Sse2.MoveMask(masked);
// existing logic
```

Reply:
Got it. So, just to double check my calculation, it doesn't matter than I changed `currentMask` to `ulong` (from `uint`) here, because we're only concerned about the right most 16 bits when ANDing with `NonAsciiDataSeenMask`. @kunalspathak